{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\section{Data management}\n",
    "\\subsection{What types of data}\n",
    "Recording data: all science is built on data\n",
    "\n",
    "Primary readouts: discrete, continuous, time-based\n",
    "\n",
    "Personally identifiable, high dimensional, high content, scale, metadata\n",
    "\n",
    "Current challenges in biomedical areas: the omics explosion, analysis, capture, data curation, search, sharing, storage, transfer, visualisation, querying, information privacy\n",
    "\n",
    "Translational science questions in a multi-omic world: \n",
    "\\begin{enumerate}\n",
    "    \\item Comparing expression profiles between healthy and diseased individuals or tissue samples\n",
    "    \\item Search multiple data sets for potential drug targets, pathways and biomarkers\n",
    "    \\item Compare data from proteomics, metabolomics and other \"omics\" studies\n",
    "    \\item Investigate correlations between genotype and phenotype in clinical trial data\n",
    "    \\item Stratify clinical data into molecular subtypes of disease and drug response\n",
    "    \\item Identify cohorts for future studies\n",
    "\\end{enumerate}\n",
    "\n",
    "\\subsection{What to capture}\n",
    "Formalising data life cycles: FAIR data principles\n",
    "\n",
    "F as findable\n",
    "\\begin{enumerate}\n",
    "    \\item F1 (Meta)data are assigned a globally unique and persisitent identifier\n",
    "    \\item F2 Data are described with rich metadata (defined by R1 below)\n",
    "    \\item Metadata cearly and explicitly include the identifier of the data it describes\n",
    "    \\item (Meta)data are registered or indexed in a searchable resource\n",
    "\\end{enumerate}\n",
    "\n",
    "A as accessible\n",
    "\\begin{enumerate}\n",
    "    \\item (meta)data are retrievable by their identifier using a standardized communications protocol\n",
    "    \\item Metadata are accessible, even when the data are no longer available\n",
    "\\end{enumerate}\n",
    "\n",
    "I as interoperable\n",
    "\\begin{enumerate}\n",
    "    \\item (meta)data use a formal, accessible, shared, and broadly applicable language for knowledge representation\n",
    "    \\item (meta)data use vocabularies that follow FAIR principles\n",
    "    \\item (meta)data include qualified references to other (meta)data\n",
    "\\end{enumerate}\n",
    "\n",
    "R as reusable\n",
    "\n",
    "\\begin{enumerate}\n",
    "    \\item(meta)data are richly desrcibed with a plurality of accurate and relevant attributes\n",
    "\\end{enumerate}\n",
    "\n",
    "It is always easier to capture too much now and delete later \n",
    "\n",
    "\\subsection{Where to capture it?}\n",
    "Paper is cheap but hard to safeguard and shared\n",
    "\n",
    "Electronic lab notebooks: labarchives\n",
    "\n",
    "Relational databases: mysql, schema-based or relational, scaling gets expensive when past millions of records or TB-PB of data, flexible querying language\n",
    "\n",
    "\\subsection{How to capture it}\n",
    "\\begin{enumerate}\n",
    "    \\item Use data standards!\n",
    "    \\item Use well-defined file formats!\n",
    "    \\item Be careful about heterogeneity\n",
    "    \\item Think carefully about metadata you need to ensure reproducibility\n",
    "    \\item Do all this before you generate a dataset to avoid much pain later!\n",
    "\\end{enumerate}\n",
    "\n",
    "Well-formatted tables from the outset\n",
    "\n",
    "Data organisation is the foundation of your research project\n",
    "\n",
    "Technical heterogeneity, data model heterogeneity, semantic heterogeneity\n",
    "\n",
    "\\section{HPC High performance computing}\n",
    "FLOPS = Floating point operations per second\n",
    "\n",
    "Next frontier: \"ExaScale\" HPC, DoE at Oak Ridge, Sunway tiahulight\n",
    "\n",
    "Spatial omics!\n",
    "\n",
    "\\subsection{Types pf HPC}\n",
    "Traditional clusters: shared, fast, access to storage, relatively cheap compute nodes\n",
    "\n",
    "Cloud-based clusters: convenience of on-demand service, expensive data transfer and high-end work\n",
    "\n",
    "OpenStack Provate Clouds: software defined cluster configuration, software define netwroking and storage, containers!\n",
    "\n",
    "\\subsection{Types of parallel computing}\n",
    "Embarrassingly parallel or distributed or serial problems: break the problem into smaller chunks, have chunks that fit efficiently iwthin each node, have appropriate access to storage for the overall problem\n",
    "\n",
    "Parallel problems: problem cannot be subdivided into independent chunks of work, problem must be broken down into parallisable and non-parallisable parts of code, optimisation of parallisable code is key and non-trivial\n",
    "\n",
    "Ahmdahl's law: the bit of code that cannot be parallelised will always be the limiting factor\n",
    "\n",
    "\\subsection{What's in an HPC cluster?}\n",
    "Compute considerations: cores (CPUs), GPGPUs, co-processors, RAM, interconnect (network), density\n",
    "\n",
    "Storage considerations: how many operations per seconds (IOPS), bandwidth (gb/s), interconnect (network), density\n",
    "\n",
    "Management nodes, power distribution, management network\n",
    "\n",
    "Job scheduler - SLURM\n",
    "\n",
    "\\section{Relational data-bases}\n",
    "Free text/unstructured\n",
    "\n",
    "CSV / Comma separated values\n",
    "\n",
    "Spreadsheets\n",
    "\n",
    "XML / Extensible Markup Language\n",
    "\n",
    "JSON / JavaScript Object Notation\n",
    "\n",
    "\\subsection{Relational model}\n",
    "All data is stored as tuples (rows) grouped into relations (tables)\n",
    "\n",
    "Relations have unique key with which each tuple can be uniquely identified\n",
    "\n",
    "A set of rules are followed when designing tables: normal forms\n",
    "\n",
    "NoSQL: is a term used to describe a series of different technologies which don't store data as relations and in general don't support SQL\n",
    "\n",
    "\\subsection{Normal forms}\n",
    "\\subsubsection{1NF}\n",
    "A single value should not contain multiple items from a single domain (no duplicates and column and row order are not important)\n",
    "\n",
    "Split single row into many rows\n",
    "\n",
    "\\subsubsection{2NF}\n",
    "No non-prime attribute is dependent on any subset of any candidate key\n",
    "\n",
    "Update anomaly: updating the MW of a protein with multiple domains will require the MW to be updated on multiple rows\n",
    "\n",
    "\\subsubsection{3NF}\n",
    "Every non-key attribute must provide a fact about the key, the whole key, and nothing but the key \n",
    "\n",
    "\\subsubsection{Normalisation: summary}\n",
    "\\begin{enumerate}\n",
    "    \\item Column and row order must be irrelevant\n",
    "    \\item Rows must not be duplicated\n",
    "    \\item Cells must have exactly one value\n",
    "    \\item Relations must describe one entitu only\n",
    "    \\item Store each piece of information only once\n",
    "\\end{enumerate}\n",
    "\n",
    "\\subsubsection{Key types}\n",
    "Surrogate primary keys are used for two reasons: consistent naming and generation of keys across all tables and where a logical key does not exist\n",
    "\n",
    "\\subsection{Creating tables}\n",
    "With design tools like \"Toad for Oracle\" and \"MySQL Workbench\".\n",
    "\n",
    "Manually by writing SQL statements\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "create table gene (geneid integer primary key, hugosymbol varchar(200) not null)\n",
    "\\end{minted}\n",
    "\n",
    "\\begin{enumerate}\n",
    "    \\item integer\n",
    "    \\item float\n",
    "    \\item double\n",
    "    \\item bool\n",
    "    \\item char(L)\n",
    "    \\item varchar(L)\n",
    "    \\item text\n",
    "    \\item date\n",
    "    \\item time\n",
    "    \\item datetime\n",
    "    \\item blob\n",
    "    \\item tinyint\n",
    "    \\item smallint\n",
    "    \\item mediumint\n",
    "    \\item bigint\n",
    "\\end{enumerate}\n",
    "\n",
    "\\subsubsection{Viewing table definition}\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "show fields from gene;\n",
    "\\end{minted}\n",
    "\n",
    "\\subsubsection{Inserting rows}\n",
    "Insert one or more values\n",
    "\n",
    "Supply values for all columns or just those which are not null\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "insert into gene (geneid) values (7157);\n",
    "\\end{minted}\n",
    "\n",
    "\\subsubsection{Select rows}\n",
    "Select statements are very flexible\n",
    "\n",
    "Fetch some or all columns in any order\n",
    "\n",
    "Fetch all rows or a subset\n",
    "\n",
    "Change the order of fetched rows\n",
    "\n",
    "Group rows and apply aggregate functions\n",
    "\n",
    "* is a placeholder for all columns\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select geneid from gene where geneid = 1234\n",
    "\\end{minted}\n",
    "\n",
    "\\subsubsection{Viewing inserted rows}\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select * from gene\n",
    "\\end{minted}\n",
    "\n",
    "\\subsection{Functions and operators}\n",
    "\\begin{enumerate}\n",
    "    \\item Equal to \"=\"\n",
    "    \\item Greater than \">\"\n",
    "    \\item Greater than or equal to \">=\"\n",
    "    \\item Less than \"<\"\n",
    "    \\item Less than or equal to \"<=\"\n",
    "    \\item Addition \"+\"\n",
    "    \\item Subtraction \"-\"\n",
    "    \\item Multiplication \"*\"\n",
    "    \\item Division \"/\"\n",
    "    \\item Contains/starts with/ends with \"like\"\n",
    "    \\item Does not (contain/start with/end with) \"not like\"\n",
    "    \\item Regular expression match \"regexp\"\n",
    "    \\item Not equal to \"!=\" \"<>\"\n",
    "    \\item Value is null \"is null\"\n",
    "    \\item Value is not null \"is not null\"\n",
    "    \\item Logical and \"and\", \"& &\"\n",
    "    \\item Local or \"or\", \"||\"\n",
    "    \\item Value is between \"between ... and ... \"\n",
    "\\end{enumerate}\n",
    "\n",
    "\\subsubsection{Aggregate functions: group by}\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select species, count(geneid) from gene_expression group by species\n",
    "\\end{minted}\n",
    "\n",
    "\\subsubsection{Aggregate functions: order by}\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select species, geneid from gene order by species asc, geneid desc\n",
    "\\end{minted}\n",
    "\n",
    "\\subsubsection{Updating rows}\n",
    "Update one or more rows\n",
    "\n",
    "Set the value for specific columns or all of them\n",
    "\n",
    "Reuse the existing to set a value\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "update gene set geneid = 123 where geneid = 2;\n",
    "\\end{minted}\n",
    "\n",
    "\\subsubsection{Deleting rows}\n",
    "Delete one row or many\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "delete from gene where geneid = 123\n",
    "\\end{minted}\n",
    "\n",
    "\\subsection{Creating relationships}\n",
    "Relational modelling of entities commonly results in one model being related to another by one of their keys\n",
    "\n",
    "Foreign keys or foreign key constraints\n",
    "\n",
    "Parent and child\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "create table transcripts (id integer primary key, geneid integer not null) foreign key (geneid) references gene (geneid)\n",
    "\\end{minted}\n",
    "\n",
    "Entity relationship diagram\n",
    "\n",
    "\\subsection{Joins}\n",
    "Using SQL you can retrieve the data stored in multiple tables in many ways\n",
    "\n",
    "\\begin{enumerate}\n",
    "    \\item \"union\" set with rows from all tables stacked vertically\n",
    "    \\item \"inner-join\" intersection of multiple tables (excludes rows that don't meet the join condition)\n",
    "    \\item \"left-outer join\" includes all rows from the left table but only those from the right table that intersect with the left table based on the join condition\n",
    "    \\item \"right-outer join\" includes all rows from the right table but only those from the left table that intersect with the right table based on the join condition\n",
    "\\end{enumerate}\n",
    "\n",
    "Intersection of multiple tables (excludes rows that do not meet the join condition)\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select a. geneid, b. transcript_id from gene a, transcript b where b. geneid = a. geneid\n",
    "\\end{minted}\n",
    "\n",
    "Outer join\n",
    "\n",
    "All protein purification records\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select a. purification_id, a.protein_sequence, b.compound_mw from protein_purification a left outer join compound b on a.compound_id = b.compound_id\n",
    "\\end{minted}\n",
    "\n",
    "Inner join\n",
    "\n",
    "Only protein purification recorded with a compound\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select a.purification_id, a.protein_sequence, b.compound_mw from protein_purification a, compound b where a.compound_id = b.compound_id\n",
    "\\end{minted}\n",
    "\n",
    "Union example\n",
    "\n",
    "Set with rows from all tables stacked vertically\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select gene_id from gene where species = \"Homo sapiens\"\n",
    "union\n",
    "select gene_id from gene where species = \"Mus musculus\"\n",
    "\\end{minted}\n",
    "\n",
    "\n",
    "\\subsection{SQL injection}\n",
    "Everyone that uses SQL as part of their work should understand how to avoid injection\n",
    "\n",
    "SQL injection is the placement of malicious code into your SQL statements\n",
    "\n",
    "SQL injection example: imagine a simple web-based form which allows you to enter your used ID and shows you all your recent orders\n",
    "\n",
    "This web-site might look-up delivieries with an SQL statement like this\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select order_no, delivery_address, product from deliveries where user_id = \" + userID\n",
    "\\end{minted}\n",
    "\n",
    "Where the used ID comes directly from the web-form\n",
    "\n",
    "Now imagine that instead of entering my user ID I enter the following into the form \"1 or 1 = 1\"\n",
    "\n",
    "Which results in this SQL being executed \n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select order_no, delivery_address, product from deliveries where user_id = 1 or 1 = 1\n",
    "\\end{minted}\n",
    "\n",
    "\"or 1=1\" ensures that I'm going to see all orders and not just mine\n",
    "\n",
    "But it gets worse, some database libraries allow for multiple SQL statements to be executed at the same time \n",
    "\n",
    "So we might be able to enter this into the from instead \"1 ; drop table deliveries\"\n",
    "\n",
    "Here we are terminating the previous SQL statement and dropping the deliveries table with a separate statement\n",
    "\n",
    "NEVER CONCATENATE NON-TRUSTED INPUT ONTO AN SQL statement!\n",
    "\n",
    "\\section{SQL}\n",
    "\\subsection{SQL syntax}\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select * from customers;\n",
    "\\end{minted}\n",
    "\n",
    "The most important commands: select, update, delete, insert into, create database, alter database, create table, alter table, drop table, create index, drop index\n",
    "\n",
    "\\subsection{SQL select statement}\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select * from customers;\n",
    "\n",
    "select CustomerName, City from Customers;\n",
    "\\end{minted}\n",
    "\n",
    "\\subsection{SQL select distinct}\n",
    "Used to return only distinct (different) values.\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select distinct country from customers;\n",
    "select count(distinct country) from customers;\n",
    "\\end{minted}\n",
    "\n",
    "\\subsection{SQL where clause}\n",
    "The where clause is used to filter records\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select * from customers where country = 'Mexico';\n",
    "select * from customers where customerID=1;\n",
    "\\end{minted}\n",
    "\n",
    "\\subsection{SQL and, or and not operators}\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select * from customers where country = 'Germany' and city = 'Berlin';\n",
    "\n",
    "select * from customers where not ountry = 'Germany';\n",
    "\\end{minted}\n",
    "\n",
    "\\subsection{SQL order by}\n",
    "Used to sort the result-set in ascending or descending order. If ordered by several columns: it orders by country, but if some rows have the same country, it orders them by name.\n",
    "\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "select * from customers order by country desc;\n",
    "\n",
    "select * from customers order by country, name;\n",
    "\\end{minted}\n",
    "\n",
    "\\subsection{SQL insert into}\n",
    "Used to insert new records in a table\n",
    "\n",
    "\n",
    "\\section{High-performace computing in python}\n",
    "\\subsection{Cython}\n",
    "Cython source files end in $.pyx$. In jupyter notebooks use the $\\%\\%cython$ magic command.\n",
    "\n",
    "\\begin{minted}[breaklines]{python}\n",
    "\\%load_ext Cython\n",
    "\\end{minted}\n",
    "\n",
    "\\begin{minted}[breaklines]{python}\n",
    "%%cython\n",
    "def norm_pyx(a, p):\n",
    "    s = 0\n",
    "    x_max = a.shape[0]\n",
    "    y_max = a.shape[1]\n",
    "    for i in range(x_max):\n",
    "        for j in range(y_max):\n",
    "            s += abs(a[i, j])**p\n",
    "    return s**(1.0/p)\n",
    "\\end{minted}\n",
    "\n",
    "Adding types: should reduce the amount of python interaction\n",
    "\n",
    "\\begin{minted}[breaklines]{python}\n",
    "%%cython -a\n",
    "cpdef double norm_pyx(a, double p):\n",
    "    cdef double s = 0.0\n",
    "    cdef Py_ssize_t x_max = a.shape[0]\n",
    "    cdef Py_ssize_t y_max = a.shape[1]\n",
    "    for i in range(x_max):\n",
    "        for j in range(y_max):\n",
    "            s += abs(a[i, j])**p\n",
    "    return s**(1.0/p)\n",
    "\\end{minted}\n",
    "\n",
    "Cython provides typed memoryviews to allow efficient access to memry buffers, such as numpy arrays. If you want to convert the memoryviews to a numpy array, use $np.asarray(memory_views)$\n",
    "\n",
    "\\begin{minted}[breaklines]{python}\n",
    "%%cython -a\n",
    "cpdef double norm_pyx(double [:,:] a, double p):\n",
    "    cdef double s = 0.0\n",
    "    cdef Py_ssize_t x_max = a.shape[0]\n",
    "    cdef Py_ssize_t y_max = a.shape[1]\n",
    "    for i in range(x_max):\n",
    "        for j in range(y_max):\n",
    "            s += abs(a[i, j])**p\n",
    "    return s**(1.0/p)\n",
    "\\end{minted}\n",
    "\n",
    "Tuning indexing further: cython uses bounds checking fro accessing arrays, divide by zero checks, and many other checks that slow down your code. Once you are confident that your code is working as expected and you don't need these checks, you can turn them off\n",
    "\n",
    "\\begin{minted}[breaklines]{python}\n",
    "%%cython -a\n",
    "from libc.math cimport abs  # Can import any libc functions you need here\n",
    "cimport cython\n",
    "\n",
    "@cython.boundscheck(False)  # Deactivate bounds checking\n",
    "@cython.wraparound(False)   # Deactivate negative indexing.\n",
    "@cython.cdivision(True)     # Deactivate normal python division checking\n",
    "\n",
    "cpdef double norm_pyx(double [:, :] a, int p):\n",
    "    cdef double s = 0\n",
    "    cdef Py_ssize_t x_max = a.shape[0]\n",
    "    cdef Py_ssize_t y_max = a.shape[1]\n",
    "    for i in range(x_max):\n",
    "        for j in range(y_max):\n",
    "            s += abs(a[i, j])**p\n",
    "    return s**(1.0/p)\n",
    "\\end{minted}\n",
    "\n",
    "Cython can also define extension types, aka cdef classes\n",
    "\n",
    "\\begin{minted}[breaklines]{python}\n",
    "class Model:\n",
    "    def __init__(self, dt):\n",
    "        self._dt = dt\n",
    "        self._y0 = 1.0\n",
    "        \n",
    "    def dydx(self, p, y):\n",
    "        return -y\n",
    "\n",
    "    def evaluate(self, p, time):\n",
    "        timesteps = int(time / self._dt)\n",
    "        y = self._y0\n",
    "        for i in range(timesteps):\n",
    "            y += self._dt * self.dydx(p, y)\n",
    "        return y\n",
    "\\end{minted}\n",
    "\n",
    "This is the equivalent cdef class\n",
    "\\begin{minted}[breaklines]{python}\n",
    "%%cython -a\n",
    "cdef class Model_pyx:                  # add cdef to convert to extension type\n",
    "    cdef double _dt                    # define C class variables as attributes using the cdef syntax\n",
    "    cdef public double _y0             # use public keyword to enable access from python\n",
    "    \n",
    "    def __cinit__(self, double dt):    # __cinit__ equivilent to C++ constructors (__init__ might not be called)\n",
    "        self._dt = dt \n",
    "        self._y0 = 1.0\n",
    "        \n",
    "    cdef double dydx(self, double p, double y): # cdef functions cannot be called from python\n",
    "        return -p*y\n",
    "\n",
    "    cpdef double evaluate(self, double p, double time): # cpdef functions *can* be called from python\n",
    "        cdef int timesteps = int(time / self._dt)\n",
    "        cdef double y = self._y0\n",
    "        cdef double tmp\n",
    "        for i in range(timesteps):\n",
    "            tmp = self.dydx(p, y)\n",
    "            y += self._dt * tmp\n",
    "        return y\n",
    "\\end{minted}\n",
    "\n",
    "Packaging Cython programs. \n",
    "\n",
    "Write a setup.py file\n",
    "\\begin{minted}[breaklines]{python}\n",
    "from distutils.core import setup\n",
    "from Cython.Build import cythonize\n",
    "\n",
    "setup(\n",
    "    ext_modules = cythonize(\"helloworld.pyx\")\n",
    ")\n",
    "\\end{minted}\n",
    "\n",
    "Build using the commandline options\n",
    "\\begin{minted}[breaklines]{bash}\n",
    "python setup.py build_ext --inplace\n",
    "\\end{minted}\n",
    "\n",
    "\n",
    "\n",
    "\\subsection{MPI}\n",
    "MPI = message passing interface\n",
    "\n",
    "Library routines designed for writing portable message-passing programs \n",
    "\n",
    "Familiarise yourself with MPI using C or Fortran: material is generally more detailed, more online tutorials, clearer learning path\n",
    "\n",
    "\\subsubsection{Message passing paradigm}\n",
    "Distributed memory programming model\n",
    "\n",
    "Independent processes running concurrently, all processes execute the same program, any data needs to be shared between processes needs to be passed explicitly\n",
    "\n",
    "Distributed computing and distributed memory\n",
    "\n",
    "The problem data is partitioned, large requirements are distributed, large requirements are distributed\n",
    "\n",
    "\\subsubsection{Advantages}\n",
    "Universality and portability: between computer systems and implementations\n",
    "\n",
    "Performance and scalability: with future core count increase, distributed memory computing is the key to high performance\n",
    "\n",
    "\\subsubsection{mpi4py}\n",
    "An MPI interface\n",
    "\n",
    "Performance not the same as for C and Fortran but what is lost in performance is gained in code development time\n",
    "\n",
    "\\begin{minted}[breaklines]{python}\n",
    "import sys\n",
    "from mpi4py import MPI\n",
    "\n",
    "def report(communicator):\n",
    "    rank = communicator.rank \n",
    "\n",
    "\\end{minted}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
